internal/kongstate/kongstate.go

```diff
@@ -89,22 +105,93 @@ func (ks *KongState) FillConsumersAndCredentials(log logrus.FieldLogger, s store
+			// this prefix in the message body to make it easier to search for this class
+			// of errors (KongConsumer credentials validation errors) in the manager logs.
+			failHeader := fmt.Sprintf("failed to provision credentials for consumer %s", consumer.Name)
+
+			// validate that the required "kongCredType" key is present in the creds.
+			credType, ok := credConfig[credTypeKey].(string)
 			if !ok {
-				log.Errorf("failed to provision credential: invalid credType: %v", credType)
+				log.Errorf("%s: secret %s provided no %s", failHeader, secret.Name, credTypeKey)
+				invalid = true
 			}
 			if !supportedCreds.Has(credType) {
-				log.Errorf("failed to provision credential: invalid credType: %v", credType)
-				continue
+				supportedCredsString := fmt.Sprintf("valid options are: %s", strings.Join(supportedCreds.List(), ","))
+				log.Errorf("%s: invalid credType: %s (%s)", failHeader, credType, supportedCredsString)
+				invalid = true
 			}
-			if len(credConfig) <= 1 { // 1 key of credType itself
-				log.Errorf("failed to provision credential: empty secret")
-				continue
+
+			// validate that the credentials actually includes configuration, not just the type
+			if len(credConfig) <= 1 {
+				log.Errorf("%s: secret %s has no data", failHeader, secret.Name)
+				continue // if this is true, there's nothing left to validate anyway
 			}
-			err = c.SetCredential(credType, credConfig, ks.Version)
-			if err != nil {
-				log.Errorf("failed to provision credential: %v", err)
-				continue
+
+			// consumer credentials can technically be configured with two (different)
+			// secrets that contain the same key. For some types there are unique
+			// contraints on the key. Here we validate the unique constraints for
+			// several types.
+			for k, v := range credConfig {
+				if k == credTypeKey {
+					continue
+				}
+
+				// TODO: filter out non-unique keys, such as "username" for "basic-auth"
+				//       See: https://github.com/kong/kong/blob/master/kong/plugins/hmac-auth/daos.lua
+
+				// validate the integrity of the secret key value data
+				value, ok := v.([]byte)
+				if !ok {
+					log.Errorf("%s: invalid credential %s: key value can't be %T", failHeader, consumer.Name, cred, v)
+					invalid = true
+				}
+
+				// check if the key has ever been seen before, if it has we need to
+				// process it to ensure deduplication.
+				//
+				// if all of the following are true (inclusively) then the key is
+				// invalid and the credentials will be dropped:
+				//
+				//  - if the credential type is the same
+				//  - if the key has a unique constraint
+				//  - if the data is duplicated
+				if metas, ok := credsDedup[k]; ok {
+					for _, meta := range metas {
+						if meta.credType == credType && bytes.Equal(meta.value, value) && supportedCredsWithUniqueConstraints.Has(credType) {
+							log.Errorf("%s: unique constaint violated: more than 1 credential secret has defined %s", failHeader, k)
+							invalid = true
```

Comment:
```suggestion
							invalid = true
							break
```
break since find break